File: src/main/java/xyz/oribuin/playerwarps/manager/DataManager.java
package xyz.oribuin.playerwarps.manager;

import com.google.gson.Gson;
import dev.rosewood.rosegarden.RosePlugin;
import dev.rosewood.rosegarden.database.DataMigration;
import dev.rosewood.rosegarden.manager.AbstractDataManager;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import xyz.oribuin.playerwarps.PlayerWarpsPlugin;
import xyz.oribuin.playerwarps.database.migration._1_CreateInitialTables;
import xyz.oribuin.playerwarps.model.Warp;
import xyz.oribuin.playerwarps.model.serializer.TextSerialized;
import xyz.oribuin.playerwarps.model.serializer.UUIDSerialized;
import xyz.oribuin.playerwarps.util.WarpUtils;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class DataManager extends AbstractDataManager {

    private static final Gson GSON = new Gson();
    private final Map<String, Warp> warps = new HashMap<>();

    public DataManager(RosePlugin rosePlugin) {
        super(rosePlugin);
    }

    @Override
    public void reload() {
        super.reload();

        this.loadWarps();
    }

    /**
     * Update the cache with the warp
     *
     * @param warp The warp to cache
     */
    public void cache(Warp warp) {
        if (warp == null) return;

        this.warps.put(warp.getId(), warp);
    }

    /**
     * Get a warp by its id from the cache
     *
     * @param id The id of the warp
     * @return The warp
     */
    public Warp get(String id) {
        return this.warps.get(id);
    }

    /**
     * Get a list of warps by their owner uuid
     *
     * @param owner The owner of the warps
     * @return The warp
     */
    public List<Warp> getOwned(UUID owner) {
        return this.warps.values()
                .stream()
                .filter(warp -> warp.getOwner().equals(owner))
                .toList();
    }

    /**
     * Update the database with the warp's data
     * In times like this we wish we were using an ORM
     *
     * @param warp The warp to update
     */
    public void save(Warp warp) {
        this.warps.put(warp.getId(), warp);

        this.async(() -> this.databaseConnector.connect(connection -> {
            String updatePrimary = "REPLACE INTO " + this.getTablePrefix() + "warps (`id`, `owner`, `owner_name`, `created`, `x`, `y`, `z`, `yaw`, `pitch`, `world`, `last_upkeep`)" +
                                   "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";

            String updateSettings = "REPLACE INTO " + this.getTablePrefix() + "settings (`id`, `display_name`, `description`, `public`, `teleport_cost`, `icon`)" +
                                    "VALUES (?, ?, ?, ?, ?, ?)";

            String updateLists = "REPLACE INTO " + this.getTablePrefix() + "lists (`id`, `banned`, `visitors`) VALUES (?, ?, ?)";

            try (
                    PreparedStatement primaryStatement = connection.prepareStatement(updatePrimary);
                    PreparedStatement settingsStatement = connection.prepareStatement(updateSettings);
                    PreparedStatement listsStatement = connection.prepareStatement(updateLists)
            ) {

                // Update the primary table with the warp's data
                primaryStatement.setString(1, warp.getId());
                primaryStatement.setString(2, warp.getOwner().toString());
                primaryStatement.setString(3, warp.getOwnerName());
                primaryStatement.setLong(4, warp.getCreationTime());
                primaryStatement.setDouble(5, warp.getPosition().getX());
                primaryStatement.setDouble(6, warp.getPosition().getY());
                primaryStatement.setDouble(7, warp.getPosition().getZ());
                primaryStatement.setFloat(8, warp.getPosition().getYaw());
                primaryStatement.setFloat(9, warp.getPosition().getPitch());
                primaryStatement.setString(10, warp.getPosition().getWorld().getName());
                primaryStatement.setLong(11, warp.getLastUpkeepTime());
                primaryStatement.executeUpdate();

                // Update the settings table with the warp's settings
                settingsStatement.setString(1, warp.getId());
                settingsStatement.setString(2, warp.getDisplayName());
                settingsStatement.setString(3, GSON.toJson(new TextSerialized(warp.getDescription())));
                settingsStatement.setBoolean(4, warp.isPublic());
                settingsStatement.setDouble(5, warp.getTeleportFee());
                settingsStatement.setBytes(6, WarpUtils.serializeItem(warp.getIcon()));
                settingsStatement.executeUpdate();

                // Update the lists table with the warp's lists
                listsStatement.setString(1, warp.getId());
                listsStatement.setString(2, GSON.toJson(new UUIDSerialized(warp.getBanned())));
                listsStatement.setString(3, GSON.toJson(new UUIDSerialized(warp.getVisitors())));
                listsStatement.executeUpdate();
            }
        }));

    }

    /**
     * Delete a warp from the database and cache
     *
     * @param id The id of the warp to delete
     */
    public void delete(String id) {
        this.warps.remove(id);

        this.async(() -> this.databaseConnector.connect(connection -> {
            String deletePrimary = "DELETE FROM " + this.getTablePrefix() + "warps WHERE id = ?";
            String deleteSettings = "DELETE FROM " + this.getTablePrefix() + "settings WHERE id = ?";
            String deleteLists = "DELETE FROM " + this.getTablePrefix() + "lists WHERE id = ?";

            try (
                    PreparedStatement primaryStatement = connection.prepareStatement(deletePrimary);
                    PreparedStatement settingsStatement = connection.prepareStatement(deleteSettings);
                    PreparedStatement listsStatement = connection.prepareStatement(deleteLists)
            ) {
                primaryStatement.setString(1, id);
                settingsStatement.setString(1, id);
                listsStatement.setString(1, id);

                primaryStatement.executeUpdate();
                settingsStatement.executeUpdate();
                listsStatement.executeUpdate();
            }
        }));
    }

    /**
     * Load all the warps from the plugin database
     * This method is called when the plugin is enabled
     */
    private void loadWarps() {
        this.warps.clear();

        this.async(() -> this.databaseConnector.connect(connection -> {

            String selectPrimary = "SELECT * FROM " + this.getTablePrefix() + "warps";
            String selectSettings = "SELECT * FROM " + this.getTablePrefix() + "settings";
            String selectLists = "SELECT * FROM " + this.getTablePrefix() + "lists";

            try (PreparedStatement statement = connection.prepareStatement(selectPrimary)) {
                ResultSet result = statement.executeQuery();

                while (result.next()) {

                    String ownerName = result.getString("owner_name");
                    Location position = new Location(
                            Bukkit.getWorld(result.getString("world")),
                            result.getDouble("x"),
                            result.getDouble("y"),
                            result.getDouble("z"),
                            result.getFloat("yaw"),
                            result.getFloat("pitch")
                    );

                    Warp warp = new Warp(result.getString("id"), UUID.fromString(result.getString("owner")), position);
                    warp.setOwnerName(ownerName);
                    warp.setCreationTime(result.getLong("created"));
                    warp.setLastUpkeepTime(result.getLong("last_upkeep"));

                    this.warps.put(warp.getId(), warp);
                }
            }

            try (PreparedStatement statement = connection.prepareStatement(selectSettings)) {
                ResultSet result = statement.executeQuery();

                while (result.next()) {
                    Warp warp = this.warps.get(result.getString("id"));
                    if (warp == null) continue;

                    warp.setDisplayName(result.getString("display_name"));
                    warp.setDescription(GSON.fromJson(result.getString("description"), TextSerialized.class).result());
                    warp.setPublic(result.getBoolean("public"));
                    warp.setTeleportFee(result.getDouble("teleport_cost"));
                    warp.setIcon(WarpUtils.deserializeItem(result.getBytes("icon")));

                    this.warps.put(warp.getId(), warp);
                }
            }

            try (PreparedStatement statement = connection.prepareStatement(selectLists)) {
                ResultSet result = statement.executeQuery();

                while (result.next()) {
                    Warp warp = this.warps.get(result.getString("id"));
                    if (warp == null) continue;

                    warp.setBanned(GSON.fromJson(result.getString("banned"), UUIDSerialized.class).result());
                    warp.setVisitors(GSON.fromJson(result.getString("visitors"), UUIDSerialized.class).result());

                    this.warps.put(warp.getId(), warp);
                }
            }
        }));
    }

    @Override
    public List<Class<? extends DataMigration>> getDataMigrations() {
        return List.of(_1_CreateInitialTables.class);
    }

    /**
     * Execute a runnable async off the main thread
     *
     * @param runnable The runnable to run async
     */
    private void async(Runnable runnable) {
        PlayerWarpsPlugin.SCHEDULER
                .scheduling()
                .asyncScheduler()
                .run(runnable);
    }

    public Map<String, Warp> getWarps() {
        return warps;
    }

}

File: src/main/kotlin/xyz/oribuin/playerwarps/manager/DataManager.kt
package xyz.oribuin.playerwarps.manager

import dev.rosewood.rosegarden.RosePlugin
import dev.rosewood.rosegarden.database.DataMigration
import dev.rosewood.rosegarden.manager.AbstractDataManager
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.entity.Player
import xyz.oribuin.playerwarps.database.migration._1_CreateInitialTables
import xyz.oribuin.playerwarps.util.ListSerializer.deserialize
import xyz.oribuin.playerwarps.util.ListSerializer.serialize
import xyz.oribuin.playerwarps.util.WarpUtils.deserializeItem
import xyz.oribuin.playerwarps.util.WarpUtils.serializeItem
import xyz.oribuin.playerwarps.warp.Warp
import java.sql.Statement
import java.util.*
import java.util.function.Consumer

class DataManager(rosePlugin: RosePlugin) : AbstractDataManager(rosePlugin) {

    val warpCache = mutableMapOf<Int, Warp>()

    override fun reload() {
        super.reload()

        // Load all warps into the cache
        this.warpCache.clear()
        this.async {
            this.databaseConnector.connect { connection ->

                // Load the primary data for the warp
                connection.prepareStatement("SELECT * FROM ${this.tablePrefix}warps").use {
                    val resultSet = it.executeQuery()
                    while (resultSet.next()) {
                        val loc = Location(
                            Bukkit.getServer().getWorld(resultSet.getString("world")),
                            resultSet.getDouble("x"),
                            resultSet.getDouble("y"),
                            resultSet.getDouble("z"),
                            resultSet.getFloat("yaw"),
                            resultSet.getFloat("pitch")
                        )

                        val warp = Warp(
                            resultSet.getInt("id"),
                            resultSet.getString("name"),
                            UUID.fromString(resultSet.getString("owner")),
                            loc
                        )

                        warp.creationTime = resultSet.getLong("creation_time")
                        warp.ownerName = resultSet.getString("owner_name")
                        this.warpCache[warp.id] = warp
                    }
                }

                // Load the settings for the warps
                connection.prepareStatement("SELECT * FROM ${this.tablePrefix}settings").use {
                    val resultSet = it.executeQuery()
                    while (resultSet.next()) {
                        val warp = this.warpCache[resultSet.getInt("id")] ?: continue

                        warp.displayName = resultSet.getString("display_name")
                        warp.description = deserialize(String::class, resultSet.getString("description")).toMutableList()
                        warp.icon = deserializeItem(resultSet.getBytes("icon"))
                        warp.isPublic = resultSet.getBoolean("public")
                        warp.teleportFee = resultSet.getDouble("teleportFee")
                    }
                }

                // Load the list of users for the warps
                connection.prepareStatement("SELECT * FROM ${this.tablePrefix}lists").use {
                    val resultSet = it.executeQuery()
                    while (resultSet.next()) {
                        val warp = this.warpCache[resultSet.getInt("id")] ?: continue

                        warp.banned = deserialize(UUID::class, resultSet.getString("banned")).toMutableList()
                        warp.visitors = deserialize(UUID::class, resultSet.getString("visitors")).toMutableList()
                        warp.likes = deserialize(UUID::class, resultSet.getString("likes")).toMutableList()
                    }
                }
            }
        }
    }

    /**
     * Create a new warp with the given name, owner, and location and then run the callback
     *
     * @param name     The name of the warp
     * @param owner    The owner of the warp
     * @param location The location of the warp
     * @param callback The callback to run after the warp is created
     */
    fun createNewWarp(name: String, owner: Player, location: Location, callback: Consumer<Warp>) {
        val warp = Warp(-1, name, owner.uniqueId, location)
        warp.ownerName = owner.name // Set the owner name
        this.async {
            this.databaseConnector.connect { connection ->

                val createPrimary = "INSERT INTO ${this.tablePrefix}warps " +
                        "(`id`, `name`, `owner`, `owner_name`, `world`, `x`, `y`, `z`, `yaw`, `pitch`, `creation_time`) " +
                        "VALUES (NULL, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"

                connection.prepareStatement(createPrimary, Statement.RETURN_GENERATED_KEYS).use { statement ->
                    statement.setString(1, warp.name)
                    statement.setString(2, warp.owner.toString())
                    statement.setString(3, owner.name)
                    statement.setString(4, location.world.name)
                    statement.setDouble(5, location.x)
                    statement.setDouble(6, location.y)
                    statement.setDouble(7, location.z)
                    statement.setFloat(8, location.yaw)
                    statement.setFloat(9, location.pitch)
                    statement.setLong(10, warp.creationTime)
                    statement.executeUpdate()
                    statement.generatedKeys.use {
                        if (it.next()) {
                            warp.id = it.getInt(1)
                            this.saveWarp(warp)
                            this.sync { callback.accept(warp) }
                        }
                    }
                }
            }
        }
    }

    /**
     * Save a warp to the database
     *
     * @param warp The warp to save
     */
    fun saveWarp(warp: Warp) {
        this.warpCache[warp.id] = warp

        this.async {
            this.databaseConnector.connect { connection ->

                // Save the primary data for the warp
                val primary = "UPDATE ${this.tablePrefix}warps SET " +
                        "`name` = ?, " +
                        "`owner` = ?, " +
                        "`owner_name` = ?, " +
                        "`world` = ?, " +
                        "`x` = ?, " +
                        "`y` = ?, " +
                        "`z` = ?, " +
                        "`yaw` = ?, " +
                        "`pitch` = ?, " +
                        "`creation_time` = ? " +
                        "WHERE `id` = ?"

                connection.prepareStatement(primary).use { statement ->
                    statement.setString(1, warp.name)
                    statement.setString(2, warp.owner.toString())
                    statement.setString(3, warp.ownerName)
                    statement.setString(4, warp.location.world.name)
                    statement.setDouble(5, warp.location.x)
                    statement.setDouble(6, warp.location.y)
                    statement.setDouble(7, warp.location.z)
                    statement.setFloat(8, warp.location.yaw)
                    statement.setFloat(9, warp.location.pitch)
                    statement.setLong(10, warp.creationTime)
                    statement.setInt(11, warp.id)
                    statement.executeUpdate()
                }


                // Save the settings for the warp
                val settings = "REPLACE INTO ${this.tablePrefix}settings " +
                        "(`id`, `display_name`, `description`, `icon`, `public`, `teleportFee`) " +
                        "VALUES (?, ?, ?, ?, ?, ?)"

                connection.prepareStatement(settings).use { statement ->
                    statement.setInt(1, warp.id)
                    statement.setString(2, warp.displayName)
                    statement.setString(3, serialize(warp.description))
                    statement.setBytes(4, serializeItem(warp.icon))
                    statement.setBoolean(5, warp.isPublic)
                    statement.setDouble(6, warp.teleportFee)
                    statement.executeUpdate()
                }


                // Save the lists for the warp
                val lists = "REPLACE INTO ${this.tablePrefix}lists " +
                        "(`id`, `banned`, `visitors`, `likes`) " +
                        "VALUES (?, ?, ?, ?)"

                connection.prepareStatement(lists).use { statement ->
                    statement.setInt(1, warp.id)
                    statement.setString(2, serialize(warp.banned))
                    statement.setString(3, serialize(warp.visitors))
                    statement.setString(4, serialize(warp.likes))
                    statement.executeUpdate()
                }
            }
        }
    }

    /**
     * Delete a warp from the database
     *
     * @param warp The warp to delete
     */
    fun deleteWarp(warp: Warp, callback: Consumer<Warp>) {
        this.warpCache.remove(warp.id)

        this.async {
            this.databaseConnector.connect { connection ->
                listOf("warps", "settings", "lists")
                    .forEach {
                        connection.prepareStatement("DELETE FROM ${this.tablePrefix}$it WHERE id = ?").use { statement ->
                            statement.setInt(1, warp.id)
                            statement.executeUpdate()
                        }
                    }
                    .let { this.sync { callback.accept(warp) } } // Run the callback on the main thread
            }
        }
    }

    override fun getDataMigrations(): List<Class<out DataMigration>> {
        return listOf(_1_CreateInitialTables::class.java)
    }


    private fun async(task: () -> Unit) =
        this.rosePlugin.server.scheduler.runTaskAsynchronously(rosePlugin, task)

    // we're using this to run a task a tick later
    private fun sync(task: () -> Unit) =
        this.rosePlugin.server.scheduler.runTask(rosePlugin, task)

}
